### PR TITLE
Use tensor cores for NT bmm

### PR DESCRIPTION
**Pull request by cpuhrsch**
Copy of internal diff.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Latest nightly prints
```
$ python benchmarks/nested/nested_bmm_bench.py --niter 100
ntensor,dtype,min_length,mean_length,max_length,runtime
4,torch.float32,106,137.25,198,0.05094399929046631
8,torch.float32,104,141.5,171,0.08773632049560547
16,torch.float32,100,141.9375,199,0.17405952453613283
32,torch.float32,100,143.5,198,0.34995201110839846
64,torch.float32,101,156.65625,199,0.8074240112304687
128,torch.float32,100,148.953125,199,1.390397491455078
256,torch.float32,100,147.72265625,200,2.522388458251953
4,torch.float16,106,132.0,149,0.04096000194549561
8,torch.float16,112,160.125,196,0.07544832229614258
16,torch.float16,100,138.4375,200,0.14750720024108888
32,torch.float16,101,139.59375,200,0.2657792091369629
64,torch.float16,100,150.1875,200,0.5373440170288086
128,torch.float16,100,145.3359375,200,1.0496205139160155
256,torch.float16,100,154.140625,200,2.1636198425292967
```
Using this PR you get
```
$ python benchmarks/nested/nested_bmm_bench.py --niter 100
ntensor,dtype,min_length,mean_length,max_length,runtime
4,torch.float32,106,137.25,198,1.2596018981933594
8,torch.float32,104,141.5,171,1.2536831665039063
16,torch.float32,100,141.9375,199,1.152583694458008
32,torch.float32,100,143.5,198,1.127413787841797
64,torch.float32,101,156.65625,199,1.171240997314453
128,torch.float32,100,148.953125,199,1.2126515197753907
256,torch.float32,100,147.72265625,200,1.2775730895996094
4,torch.float16,106,132.0,149,1.0807603454589845
8,torch.float16,112,160.125,196,1.099694061279297
16,torch.float16,100,138.4375,200,1.0823680114746095
32,torch.float16,101,139.59375,200,1.0837811279296874
64,torch.float16,100,150.1875,200,1.136875534057617
128,torch.float16,100,145.3359375,200,1.2546662139892577
256,torch.float16,100,154.140625,200,1.2935475158691405
```

There are clearly some issues with overhead, but for the large batch case you get a rough 2x performance improvement. I'll do more investigation in follow up. The nightly uses a plain for loop. Will add more tests to this PR before merging.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 1]
```diff
@@ -821,6 +821,38 @@ Tensor bmm_nested_cuda(const Tensor& self, const Tensor& mat2) {
...
+      for (int64_t i = 0; i < ntensors; i++) {
+        aptr.push_back(self_buffer.data_ptr<c10::Half>() + a_offsets[i]);
+        bptr.push_back(mat2_buffer.data_ptr<c10::Half>() + b_offsets[i]);
+        dptr.push_back(out_buffer.data_ptr<c10::Half>() + output_offsets[i]);
+      }
+      gemm_grouped_cuda_internal<c10::Half>(
+          lda, ldb, ldd, aptr, bptr, dptr, gemm_sizes, ntensors, device);
+      return output;
+    }
+>>>>>>> b8b1d7be24a29d9b20b25c0dd5273a499af07097
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu]: I think this might be rebase fluff

**Review by drisspg — APPROVED** [review]
Nit: about the fluff but otherwise the list of pointers is cool

**Pull request closed without merging**